src/lsp/cobol_parser/expect/gen_expect.ml

```diff
@@ -0,0 +1,421 @@
+     [@default] attribute on a nullable nonterminal in the grammar *)
+  | T _ -> raise Not_found
+  | N nt -> begin match nt with
+|};
+  Nonterminal.fold (fun nt acc -> begin
+    match best_guess_default_value nt with
+      | None -> acc
+      | Some default -> ([nt], default)::acc
+  end) []
+  |> pp_match_cases ppf
+    Fmt.(using Nonterminal.mangled_name (any "N_" ++ string))
+    Fmt.string
+    "raise Not_found end"
+
+module DEBUG = struct
+
+let emit_firsts ppf =
+  Nonterminal.iter (fun nonterm ->
+    let kind = if Nonterminal.kind nonterm  == `START then "start" else "" in
+    let nullable = if Nonterminal.nullable nonterm then "nullable" else "" in
+    Fmt.(
+      pf ppf "-%a- %s %s\n\t[%a]\n\n"
+      Print.nonterminal nonterm
+      nullable kind
+      (pp_list Print.terminal) (Nonterminal.first nonterm)))
+
+let emit_state_productions ppf =
+  Lr1.iter (fun lr1 -> begin
+    let lr0 = Lr1.lr0 lr1 in
+    let items = Lr0.items lr0 in
+    Fmt.pf ppf "\nState lr1-%d- lr0(%d):\n%a"
+      (Lr1.to_int lr1) (Lr0.to_int lr0)
+      Fmt.(option ~none:nop (any "DEFAULT-PROD: " ++ Print.production))
+      (Lr1.default_reduction lr1);
+    Print.itemset ppf items;
+    Fmt.pf ppf "TRANSITIONS [%a]\n"
+      (Fmt.list ~sep:(Fmt.any "; ") (fun ppf (s, i) ->
+        Fmt.pf ppf "%a(%d)" Print.symbol s (Lr1.to_int i))) (Lr1.transitions lr1);
+    Fmt.pf ppf "REDUCTIONS [%a]\n"
+      Fmt.(list ~sep:(any "; ") (using fst Print.terminal)) (Lr1.get_reductions lr1);
+  end)
+
+let emit_temp ppf =
+  Lr1.iter (fun lr1 -> begin
+    let lr0 = Lr1.lr0 lr1 in
+    let items = Lr0.items lr0 in
+    let r1 = reducable_productions lr1 in
+    let r2 = reducable_productions lr1 in
+    if List.equal Production.equal r1 r2 then () else begin
+    Fmt.pf ppf "\nState lr1-%d- lr0(%d):\n%a"
+      (Lr1.to_int lr1) (Lr0.to_int lr0)
+      Fmt.(option ~none:nop (any "DEFAULT-PROD: " ++ Print.production))
+      (Lr1.default_reduction lr1);
+    Print.itemset ppf items;
+    Fmt.string ppf "R1:\n";
+    Print.itemset ppf (List.map (fun p -> (p, 0)) r1);
+    Fmt.string ppf "R2:\n";
+    Print.itemset ppf (List.map (fun p -> (p, 0)) r2) end
+  end)
+
+let emit_nullable_unrecoverable ppf =
+  let nt_without_default = Nonterminal.fold (fun nt acc -> begin
+    if is_nullable_without_recovery nt
+    && best_guess_default_value nt == None
+    then nt::acc else acc
+  end) [] in
+  if nt_without_default == []
+  then Fmt.string ppf "All nullable nonterminals have a recovery/default/guessed-default"
+  else List.iter (fun nt ->
+    Fmt.pf ppf "%s  (%s)\n"
+      (Nonterminal.mangled_name nt)
+      (Option.value ~default:"no typ" @@ Nonterminal.typ nt))
+  nt_without_default
+
+end
+
+
+let () =
+  let ppf = Fmt.stdout in
+  Fmt.pf ppf
+    "(* Caution: this file was automatically generated from %s; do not edit *)\
+     \nmodule NEL = Cobol_common.Basics.NEL\
+     \nmodule Menhir = Grammar.MenhirInterpreter\
+     \nopen Grammar_tokens@\n\n"
```

Comment:
That might be where `!external_tokens` is needed. You could also require a mandatory `--nel-module` argument for `NEL` (in case that script is re-used somewhere else, for another parser).